### ISSUE TITLE
A BUG in the sweep pipeline ?

### ISSUE THREAD

**Issue by zhangleisdau** [jclev-uga/SWEEP#7]
Hi, Dr. Josh. 
      Recently, i am focus on calling SNPs in a polyploid species.After manually check the original code, i found there may be a bug in the "Haplotype.py" script. The possible mistake code in the "Haplotype.py" script were shown below: 
![123](https://user-images.githubusercontent.com/7265307/47955613-b1464280-dfd5-11e8-80cc-ea012b243456.png)
      In line 74,75 and 83,84 (the two red box in the picture); these two "determine statements" should not be in the "for statement", they should be placed after the "for statement syntax" (as shown below).
![456](https://user-images.githubusercontent.com/7265307/47955616-b3a89c80-dfd5-11e8-86a1-20fab6312b0c.png)
     Am I right about this? Sorry for boring you. Thank you very much. I appreciate your reply!
Best! 
                                                                                                                         Lei zhang

**Comment by jclev-uga** [comment]
Hi Lei


    First - if you are interested please use this updated version: https://github.com/w-korani/SWEEP_GOLD

[https://avatars1.githubusercontent.com/u/21265433?s=400&v=4]<https://github.com/w-korani/SWEEP_GOLD>

w-korani/SWEEP_GOLD<https://github.com/w-korani/SWEEP_GOLD>
github.com
An updated version of https://github.com/jclev-uga/SWEEP - w-korani/SWEEP_GOLD




   It is vastly superior.  We used this pipeline on a new array and achieved much greater results.  You can read about it in the newer paper out this year in Frontiers in Plant Science special issue on polyploids.


    About your question.  I am not sure.  It works exactly as I intended.  However, I can see that if you are using many samples this may cause some false positives.  This was so long ago, I am having trouble recapturing my mind when I wrote it.


    Thanks for your interest and perhaps you caught something.  I may make that change although I need to think about it some more.  There may be a reason why I did it that way - to save loops through or something.


Josh

________________________________

**Comment by zhangleisdau** [comment]
Hi Josh:
    Thank you for your quick reply!  In the "Haplotype.py" script, homologous SNPs was selected by
anchoring them with the homeologous SNPs. Accroding your explanation in the paper, **the homeologous SNPs should be heterozygous in all genotypes**. So from the line 71 to 73 (the "for statement"),  the loops shpuld screen all the samples to check if they are all heterozygous ("normalized Phred scale likelihoods of the genotypes" should be all 'heterozygous', which represented by "2" in the script). And then, followed by the line 74 to 75. Do I Understand It Correctly? Or something wrong with me?
    I am working on a hexaploid species (sweet potato), and there are 88 genotypes in the population. All of these genotypes were RNA sequenced. And the doploid genome of the wild ancestor for sweet potato was released, recently. I found that you have many softwares to call SNP from polyploid genome, such as SWEEP, SWEEP_GOLD, HAPLOSWEEP and SNP-ML. I was confused to select the best SNP calling software for the sweet potato. Could you please recommand me the appropriate software for the sweet potato?
     Thank you very much! Looking forward to your relpy!
Best!
                                                                                                                                                Lei zhang

**Comment by jclev-uga** [comment]
Yes I think you are right actually.  It may be a bug.  However when I check the results it is performing as intended so I am still working it out.


My recommendation is to use SWEEP_GOLD.  It is vastly superior.  However, we also developed SNPML using machine learning as an additional filter fro SWEEP.  One of the limitations of the original pipeline was my programming skills.  I used the likelihoods for the filtering and when I would go and look at the actual sequences piling up at a site I found that the likelihoods could actually misrepresent the data.  SNPML used the array information to train good models.  We used that to design a second array and the false positive rate was very low.  SWEEP_GOLD was still better.


So use SWEEP_GOLD OR you can use SWEEP first and the filter those results with SNPML.  Either path should give you fine results.  Again thanks for your interest and please let me know if it was successful for you.  If it does work well for you then spread the word!


We struggled so much in peanut before developing these pipelines.  Since then are genotyping has been incredibly successful.   I am excited to share and hope that our success can help others as well.


Josh

________________________________
